packages/mui-material/src/styles/ThemeProviderWithVars.test.js

```diff
@@ -408,4 +408,31 @@ describe('[Material UI] ThemeProviderWithVars', () => {
       fireEvent.click(screen.getByText('Dark'));
     }).not.toErrorDev();
   });
+
+  it('theme should remain the same when ThemeProvider rerenders', () => {
+    const theme = createTheme({ cssVariables: true });
+
+    function Inner() {
+      const upperTheme = useTheme();
+      const [count, setCount] = React.useState(0);
+      React.useEffect(() => {
+        setCount((prev) => prev + 1);
+      }, [upperTheme]);
+      return <span data-testid="inner">{count}</span>;
+    }
+    function App() {
+      const [count, setCount] = React.useState(0);
+      return (
+        <ThemeProvider theme={theme}>
+          <button onClick={() => setCount(count + 1)}>rerender</button>
+          <Inner />
+        </ThemeProvider>
+      );
+    }
+    render(<App />);
+
+    fireEvent.click(screen.getByRole('button'));
+
+    expect(screen.getByTestId('inner')).to.have.text('2'); // due to double rendering
+  });
```

Comment:
Should `theme` and `upperTheme` be referentially equal? I was about to suggest that we could use another approach for the test: instead of counting renders, check that `theme === upperTheme`, but I found that they're not referentially equal. Not sure if it's possible to do, just wondering if that's something we should aim for.

```jsx
it('theme should remain the same when ThemeProvider rerenders', () => {
  const theme = createTheme({ cssVariables: true });

  function Inner() {
    const upperTheme = useTheme();
    return <span data-testid="inner">{(theme === upperTheme).toString()}</span>;
  }
  function App() {
    return (
      <ThemeProvider theme={theme}>
        <Inner />
      </ThemeProvider>
    );
  }
  render(<App />);

  expect(screen.getByTestId('inner')).to.have.text('true');
});
```

Reply:
Good point but `const theme = createTheme({ cssVariables: true });` is not equal to `const upperTheme = useTheme();` because it's recalculated within `ThemeProvider`.

I could write a different test using `React.useRef` to compare the previous and the incoming theme so that it's not rely on count.

Reply:
Sounds good!

Reply:
Updated.